runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -11963,3 +11986,295 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   {
+   bool dynLitPoolLoad = false;
+   TR::Register* resultReg = TR::TreeEvaluator::checkAndAllocateReferenceRegister(node, cg, dynLitPoolLoad);
+
+   if (TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier())
+      {
+      resultReg = J9::Z::TreeEvaluator::generateSoftwareReadBarrier(node, cg, resultReg, NULL,
+                                                                    cg->comp()->useCompressedPointers(),
+                                                                    true);
+      }
+   else
+      {
+      TR::MemoryReference* loadMemRef = generateS390MemoryReference(node, cg);
+      TR::TreeEvaluator::checkAndSetMemRefDataSnippetRelocationType(node, cg, loadMemRef);
+      TR::InstOpCode::Mnemonic loadOp = cg->comp()->useCompressedPointers() ? TR::InstOpCode::LLGFSG : TR::InstOpCode::LGG;
+      generateRXInstruction(cg, loadOp, node, resultReg, loadMemRef);
+      node->setRegister(resultReg);
+      }
+
+   return resultReg;
+   }
+
+/**
+ * \details
+ * Resolved and unresolved reference field load get two slightly different sequences.
+ *
+ * Resolved reference fields load sequence for -XnocompressedRefs:
+ * \verbatim
+ *
+ * Label:   startICF
+ * LG   R_obj, Ref_field_MemRef
+ *
+ * // range check with implicit CS cycle check
+ * CLG  R_obj, EvacuateBase(R_vmthread)
+ * BRC  COND_BL, doneLabel
+ * CLG  R_obj, EvacuateTop(R_vmthread)
+ * BRC  COND_BH, doneLabel
+ *
+ * LAY  R_addr, Ref_field_MemRef
+ * STG  R_addr, floatTemp1(R_vmthread)
+ * BRC  call_snippet_to_jitReadBarrier (may use trampoline)
+ *
+ * Label: jitReadBarrier return label
+ * // reload evacuated reference
+ * LG   R_obj, 0(R_addr)
+ *
+ * doneLabel: endICF
+ * \endverbatim
+ *
+ *
+ * Unresolved reference fields load sequence for -XnocompressedRefs:
+ * \verbatim
+ *
+ * Label:   startICF
+ * LAY  R_addr, Ref_field_MemRef
+ * LG   R_obj, 0(R_addr)
+ *
+ * // range check with implicit CS cycle check
+ * CLG  R_obj, EvacuateBase(R_vmthread)
+ * BRC  COND_BL, doneLabel
+ * CLG  R_obj, EvacuateTop(R_vmthread)
+ * BRC  COND_BH, doneLabel
+ *
+ * STG  R_addr, floatTemp1(R_vmthread)
+ * BRC  call_snippet_to_jitReadBarrier (may use trampoline)
+ *
+ * Label: jitReadBarrier return label
+ * // reload evacuated reference
+ * LG   R_obj, 0(R_addr)
+ *
+ * doneLabel: endICF
+ * \endverbatim
+ *
+ * If compressed pointer is enabled, the LG instructions above are replaced by LLGF+SLLG.
+ */
+TR::Register *
+J9::Z::TreeEvaluator::generateSoftwareReadBarrier(TR::Node* node,
+                                                  TR::CodeGenerator* cg,
+                                                  TR::Register* resultReg,
+                                                  TR::MemoryReference* memRef,
+                                                  const bool isCompressedRef,
+                                                  const bool hasReturnReg)
+   {
+#ifndef OMR_GC_CONCURRENT_SCAVENGER
```

Comment:
Do we need this `#ifdef`? Will it not always be true for Z platforms in OpenJ9?

Reply:
This `ifdef` is no longer in my new commit cfca0d4266.